# Logistic regression baseline classifier
-  feature engineering, and how they help improve model performance. 
- compare the accuracy score of Logistic Regression models on the dataset before and after feature engineering by comparing test labels with the predicted values of the target variable Loan Status

In [2]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split
import pandas as pd

In [3]:
loan_data = pd.read_csv('data/LoanTrainingSetScaledAndOHE.csv')
loan_data.head()

Loan Status  Current Loan Amount  Credit Score  Years in current job  \
0            1                 2.00          0.84                 -0.48   
1            1                 0.32          0.49                  0.91   
2            1                -1.06          1.10                 -1.31   
3            1                -0.87         -0.63                  1.19   
4            0                 1.06         -1.31                 -0.20   

   Years of Credit History  Months since last delinquent  \
0                     0.84                          0.29   
1                    -1.97                          0.77   
2                    -1.46                         -0.48   
3                     2.00                          1.47   
4                     0.07                          1.53   

   Number of Open Accounts  Number of Credit Problems  Current Credit Balance  \
0                     0.18                      -0.34                    0.82   
1                     1.79                      -0.34                    0.56   
2                    -0.83                      -0.34                   -0.70   
3                     0.58                      -0.34                   -0.08   
4                    -0.83                      -0.34                   -0.21   

   Maximum Open Credit  ...  Purpose_EducationalExpenses  \
0                 0.19  ...                            0   
1                 1.05  ...                            0   
2                -0.55  ...                            0   
3                -0.22  ...                            0   
4                -0.38  ...                            0   

   Purpose_HomeImprovements  Purpose_HomePurchase  Purpose_MajorPurchase  \
0                         0                     0                      0   
1                         0                     0                      0   
2                         0                     0                      0   
3                         1                     0                      0   
4                         0                     0                      0   

   Purpose_MedicalBills  Purpose_Moving  Purpose_Other  \
0                     0               0              0   
1                     0               0              0   
2                     0               0              1   
3                     0               0              0   
4                     0               0              0   

   Purpose_RenewableEnergyPurchase  Purpose_Vacation  Purpose_Wedding  
0                                0                 0                0  
1                                0                 0                0  
2                                0                 0                0  
3                                0                 0                0  
4                                0                 0                0  

[5 rows x 31 columns]

In [6]:
loan_data.columns

Index(['Loan Status', 'Current Loan Amount', 'Credit Score',
       'Years in current job', 'Years of Credit History',
       'Months since last delinquent', 'Number of Open Accounts',
       'Number of Credit Problems', 'Current Credit Balance',
       'Maximum Open Credit', 'Bankruptcies', 'Tax Liens', 'DTI Ratio',
       'Term_Long Term', 'Term_Short Term', 'Home Ownership_Mortgage',
       'Home Ownership_Own Home', 'Home Ownership_Rent',
       'Purpose_BusinessLoan', 'Purpose_CarPurchase',
       'Purpose_DebtConsolidation', 'Purpose_EducationalExpenses',
       'Purpose_HomeImprovements', 'Purpose_HomePurchase',
       'Purpose_MajorPurchase', 'Purpose_MedicalBills', 'Purpose_Moving',
       'Purpose_Other', 'Purpose_RenewableEnergyPurchase', 'Purpose_Vacation',
       'Purpose_Wedding'],
      dtype='object')

In [4]:
# Create X matrix and y array
X = loan_data.drop("Loan Status", axis=1)
y = loan_data["Loan Status"]

# Train/test split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=123)

# Instantiate
logistic = LogisticRegression()

# Fit
logistic.fit(X_train, y_train)

# Predict and print accuracy
print(accuracy_score(y_true=y_test, y_pred=logistic.predict(X_test)))

C:\Users\Shubham\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:433: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)


0.7209333333333333


- **Task** : Convert Annual Income to monthly, and derive the ratio of Monthly Debt to monthly_income and store it in dti_ratio

In [5]:
# Convert income: monthly_income
monthly_income = loan_data["Annual Income"]/12

# Make engineered feature, remove features used
loan_data["dti_ratio"] = loan_data["Monthly Debt"]/monthly_income * 100
loan_data = loan_data.drop(["Monthly Debt","Annual Income"], axis=1)

KeyError: 'Annual Income'

- Convert the target variable to numerical values and replace categorical features with dummy values.

In [ ]:
# Create dti_ratio variable
monthly_income = loan_data["Annual Income"]/12
loan_data["dti_ratio"] = loan_data["Monthly Debt"]/monthly_income * 100
loan_data = loan_data.drop(["Monthly Debt","Annual Income"], axis=1)

# Replace target variable levels
loan_data["Loan Status"] = loan_data["Loan Status"].replace({'Fully Paid': 0, 
                                            'Charged Off': 1})

# One-hot encode categorical variables
loan_data = pd.get_dummies(data=loan_data)

# Print
print(loan_data.head())

- Fit and predict a Logistic Regression on loans_dti and evaluate the trained model's accuracy score.

In [ ]:
# Create X matrix and y array
X = loans_dti.drop("Loan Status", axis=1)
y = loans_dti["Loan Status"]

# Train/test split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=123)

# Instantiate
logistic_dti = LogisticRegression()

# Fit
logistic_dti.fit(X_train, y_train)

# Predict and print accuracy
print(accuracy_score(y_true=y_test, y_pred=logistic_dti.predict(X_test)))

#### Engineering features is a great way to increase the predictive accuracy of your models. You've just seen how the accuracy went from a baseline LogisticRegression model with 71% accuracy to 72% using a feature you engineered